lib/ansible/plugins/inventory/vmware_host_inventory.py

```diff
@@ -0,0 +1,514 @@
+                valid = True
+
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        """
+        Parses the inventory file
+        """
+
+        if not HAS_REQUESTS:
+            raise AnsibleParserError('Please install "requests" Python module as this is required'
+                                     ' for VMware Guest dynamic inventory plugin.')
+        elif not HAS_PYVMOMI:
+            raise AnsibleParserError('Please install "PyVmomi" Python module as this is required'
+                                     ' for VMware Guest dynamic inventory plugin.')
+        if HAS_REQUESTS:
+            # Pyvmomi 5.5 and onwards requires requests 2.3
+            # https://github.com/vmware/pyvmomi/blob/master/requirements.txt
+            required_version = (2, 3)
+            requests_version = requests.__version__.split(".")[:2]
+            try:
+                requests_major_minor = tuple(map(int, requests_version))
+            except ValueError:
+                raise AnsibleParserError("Failed to parse 'requests' library version.")
+
+            if requests_major_minor < required_version:
+                raise AnsibleParserError("'requests' library version should"
+                                         " be >= %s, found: %s." % (".".join([str(w) for w in required_version]),
+                                                                    requests.__version__))
+
+        super(InventoryModule, self).parse(inventory, loader, path, cache=cache)
+
+        cache_key = self.get_cache_key(path)
+
+        config_data = self._read_config_data(path)
+
+        source_data = None
+        if cache:
+            cache = self.get_option('cache')
+
+        update_cache = False
+        if cache:
+            try:
+                source_data = self.cache.get(cache_key)
+            except KeyError:
+                update_cache = True
+
+        # set _options from config data
+        self._consume_options(config_data)
+
+        self._set_credentials()
+        self.content = self._login()
+
+        try:
+            api_type = self.content.about.apiType
+        except (vmodl.RuntimeFault, vim.fault.VimFault) as exc:
+            raise AnsibleParserError("Failed to get status of vCenter server : %s" % exc.msg)
+
+        if api_type != 'VirtualCenter':
+            raise AnsibleParserError("Please run vmware_host_inventory against vCenter server")
+
+        if self.with_tags:
+            self.rest_content = self._login_vapi()
+
+        using_current_cache = cache and not update_cache
+        cacheable_results = self._populate_from_source(source_data, using_current_cache)
+
+        if update_cache:
+            self.cache.set(cache_key, cacheable_results)
+
+    def _populate_from_cache(self, source_data):
+        """
+        Populate inventory from cache
+        """
+        hostvars = source_data.pop('_meta', {}).get('hostvars', {})
```

Comment:
Have you tested caching with a file-backed cache plugin? I'm code-reviewing since I can't run this so forgive me if I'm mistaken, but I don't see where you've formatted the cache results with '_meta' and 'hostvars' when caching the data.
If you add
```
cache: True
cache_plugin: jsonfile
cache_connection: /tmp/inventorypath
```
to your config file and use ansible-inventory a couple times you should be able to see if this works as expected.

aws_ec2, which you may have based the caching off, initially formatted results before putting them in the cache. The cacheable results would end up looking something like:
```
{
    'aws_ec2_group_containing_hosts_here': [
        'hostname1',
        'hostname2',
        etc],
    '_meta':
        # hostvars if the host has any associated; hostvars were retrieved by self.inventory.get_host('hostname1')
        'hostvars': {
            'hostname1': {'key1': 'val1', 'key2': 'val2'},
            'hostname2': {'key1': 'val1', 'key2': 'val2'},
            etc
        }
}
```
I did that so the caching from the old ec2.py contrib script and the new plugin looked identical, but a few months ago this was simplified so the plugin just caches the raw data from the API call so the populating from cache or source is the same. I'm not sure if that would apply here well also, but thought I'd mention it.